src/app/shared/table-of-contents/table-of-contents.ts

```diff
@@ -0,0 +1,177 @@
+  active: boolean;
+
+  /* name of the anchor */
+  name: string;
+
+  /* top offset px of the anchor */
+  top: number;
+}
+
+@Component({
+  selector: 'table-of-contents',
+  styleUrls: ['./table-of-contents.scss'],
+  template: `
+    <div *ngIf="links?.length" class="docs-toc-container">
+      <div class="docs-toc-heading">Contents</div>
+      <nav>
+        <a [href]="_rootUrl + '#' + link.id"
+          *ngFor="let link of links; let i = index"
+          class="docs-level-{{link.type}}"
+          [class.active]="link.active">
+          {{link.name}}
+        </a>
+      </nav>
+    </div>
+  `
+})
+export class TableOfContents implements OnInit, AfterViewInit {
+
+  @Input() links: Link[] = [];
+  @Input() container: string;
+  @Input() headerSelectors = '.docs-markdown-h3,.docs-markdown-h4';
+
+  _rootUrl: string;
+  private _scrollContainer: any;
+  private _scrollSubscription: Subscription;
+  private _routeSubscription: Subscription;
+  private _fragmentObserver: MutationObserver;
+  private _headerObserver: MutationObserver;
+
+  private get scrollOffset(): number {
+    const {top} = this._element.nativeElement.getBoundingClientRect();
+    if (typeof this._scrollContainer.scrollTop !== 'undefined') {
+      return this._scrollContainer.scrollTop + top;
+    } else if (typeof this._scrollContainer.pageYOffset !== 'undefined') {
+      return this._scrollContainer.pageYOffset + top;
+    }
+  }
+
+  constructor(private _router: Router,
+              private _route: ActivatedRoute,
+              private _element: ElementRef,
+              @Inject(DOCUMENT) private _document: any) {
+
+    this._routeSubscription = this._router.events.subscribe((event: Event) => {
+      if (event instanceof NavigationEnd) {
+        const rootUrl = _router.url.split('#')[0];
+        if (rootUrl !== this._rootUrl) {
+          this.links.splice(0, this.links.length);
+          this.createLinks();
+          this._rootUrl = rootUrl;
+        }
+      }
+    });
+  }
+
+  ngOnInit(): void {
+    this._scrollContainer = this.container ?
+      this._document.querySelectorAll(this.container)[0] : window;
+
+    this._scrollSubscription = Observable
+      .fromEvent(this._scrollContainer, 'scroll')
+      .debounceTime(10)
+      .subscribe(this.onScroll.bind(this));
+  }
+
+  ngAfterViewInit(): void {
+    this._route.fragment.subscribe(fragment => {
+      // ensure we don't create memory leaks
+      if (this._fragmentObserver) {
+        this._fragmentObserver.disconnect();
+      }
+
+      // create a observer to ensure the element is loaded
+      // into view before trying to query and scroll to it
+      this._fragmentObserver = new MutationObserver(() => {
+        const target = document.getElementById(fragment);
+        if (target) {
+          target.scrollIntoView();
+          this._fragmentObserver.disconnect();
+          this._fragmentObserver = null;
+        }
+      });
+
+      this._fragmentObserver.observe(document.body, {
```

Comment:
This is attached to the body because it could be the document viewer or the guide viewer. In an effort, to keep it generic i attached it to the body. I could add an input if you like to pick a specific element to listen on. Let me know either way.